src/operator/matrix_op-inl.h

```diff
@@ -599,24 +601,74 @@ struct SliceParam : public dmlc::Parameter<SliceParam> {
   int begin;
   int end;
   DMLC_DECLARE_PARAMETER(SliceParam) {
-    DMLC_DECLARE_FIELD(axis).set_lower_bound(0)
-      .describe("The axis to be sliced");
-    DMLC_DECLARE_FIELD(begin).set_lower_bound(0)
-      .describe("The beginning index to be sliced");
-    DMLC_DECLARE_FIELD(end).set_lower_bound(0)
-      .describe("The end index to be sliced");
+    DMLC_DECLARE_FIELD(axis).set_lower_bound(0).describe(
+        "The axis to be sliced");
+    DMLC_DECLARE_FIELD(begin).describe(
+        "The beginning index to be sliced. Following numpy slice semantics, "
+        "None is equivalent to begin=0, -1 is the second last value, etc. For "
+        "example, if the axis dimension is of size 5, then begin=-1 is "
+        "equivalent to end=4, begin=-2 equivalent to end=3, etc.");
+    DMLC_DECLARE_FIELD(end).describe(
+        "The end index to be sliced. Following numpy slice semantics, None is "
+        "equivalent to slicing up to the last value, -1 the second last "
+        "value, etc. For example, if the axis dimension is of size 5, then "
+        "end=None is equivalent to end=5, end=-1 equivalent to end=4, etc. ");
   }
 };
 
-inline TShape SliceShape(const TShape& ishape,
-                         const EnvArguments& env) {
+inline void parseSliceSpec(const EnvArguments& env, SliceParam* param,
+                    const TShape& input) {
+  std::map<std::string, std::string> store;
+  for (auto i : env.kwargs) {
+    store[i.first] = i.second;
+  }
+
+  try {
+    // get axis
+    CHECK_GT(store.count("axis"), 0)
+        << "need to specify the non-optional axis argument";
+    param->axis = std::stoi(store["axis"]);
+    CHECK_GE(param->axis, 0) << "axis argument cannot be negative";
+
+    // get axis size
+    int axis_size = static_cast<int>(input[param->axis]);
+
+    // get begin + end
+    if ((store.count("begin") == 0) || (store["begin"] == "None"))
```

Comment:
No, a correct approach. Because the parameter structure is a plain data structure as it get declared. A correct approach will be define a class option<T>, and in the class definition

struct MyParam {
    option<int> begin, end;
}

and support parameter setting for option values